java-squid/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java

```diff
@@ -62,91 +61,101 @@
-      "java.io.CharArrayReader",
-      "java.io.CharArrayWriter",
-      "com.sun.org.apache.xml.internal.security.utils.UnsyncByteArrayOutputStream"
-    };
+  private static final String JAVA_IO_CLOSEABLE = "java.io.Closeable";
+  private static final String[] IGNORED_CLOSEABLE_SUBTYPES = {
+    "java.io.ByteArrayOutputStream",
+    "java.io.ByteArrayInputStream",
+    "java.io.StringReader",
+    "java.io.StringWriter",
+    "java.io.CharArrayReader",
+    "java.io.CharArrayWriter",
+    "com.sun.org.apache.xml.internal.security.utils.UnsyncByteArrayOutputStream"
+  };
+
+  static boolean needsClosing(Type type) {
+    for (String ignoredTypes : IGNORED_CLOSEABLE_SUBTYPES) {
+      if (type.is(ignoredTypes)) {
+        return false;
+      }
+    }
+    return isCloseable(type);
+  }
 
-    ProgramState programState;
+  static boolean isCloseable(Type type) {
+    return type.isSubtypeOf(JAVA_IO_CLOSEABLE);
+  }
 
-    protected SingleTreeNodeVisitor(ProgramState programState) {
-      this.programState = programState;
+  static boolean isOpeningResource(NewClassTree syntaxNode) {
+    if (isWithinTryHeader(syntaxNode)) {
+      return false;
     }
+    final Type type = syntaxNode.symbolType();
+    return needsClosing(type);
+  }
 
-    @Override
-    protected void scan(Tree tree) {
-      // Cut recursive processing
+  static boolean isWithinTryHeader(NewClassTree syntaxNode) {
+    final Tree parent = syntaxNode.parent();
+    if (parent != null) {
+      final Tree grandParent = parent.parent();
+      if (grandParent != null) {
+        final Tree greatGrandParent = grandParent.parent();
+        return greatGrandParent != null && greatGrandParent.is(Tree.Kind.TRY_STATEMENT);
+      }
     }
+    return false;
+  }
 
-    @Override
-    protected void scan(List<? extends Tree> trees) {
-      // Cut recursive processing
-    }
+  private static class PostStatementVisitor extends CheckerTreeNodeVisitor {
 
-    @Override
-    protected void scan(ListTree<? extends Tree> listTree) {
-      // Cut recursive processing
+    public PostStatementVisitor(CheckerContext context) {
+      super(context.getState());
     }
 
-    protected static boolean needsClosing(Type type) {
-      for (String ignoredTypes : IGNORED_CLOSEABLE_SUBTYPES) {
-        if (type.is(ignoredTypes)) {
-          return false;
+    @Override
+    public void visitNewClass(NewClassTree syntaxNode) {
+      if (isOpeningResource(syntaxNode)) {
+        final SymbolicValue instanceValue = programState.peekValue();
+        if (!(instanceValue instanceof ResourceWrapperSymbolicValue)) {
+          programState = programState.addConstraint(instanceValue, new ObjectConstraint(syntaxNode, States.OPENED));
         }
       }
-      return isCloseable(type);
     }
+  }
 
-    protected static boolean isCloseable(Type type) {
-      return type.isSubtypeOf(JAVA_IO_CLOSEABLE);
-    }
+  public static class ResourceWrapperSymbolicValue extends SymbolicValue {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Override this superclass' "equals" method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2160)